server/image_list.go

```diff
@@ -11,15 +11,31 @@ import (
 func (s *Server) ListImages(ctx context.Context, req *types.ListImagesRequest) (*types.ListImagesResponse, error) {
 	_, span := log.StartSpan(ctx)
 	defer span.End()
-	filter := ""
-	reqFilter := req.Filter
-	if reqFilter != nil {
-		filterImage := reqFilter.Image
-		if filterImage != nil {
-			filter = filterImage.Image
+
+	if reqFilter := req.Filter; reqFilter != nil {
+		if filterImage := reqFilter.Image; filterImage != nil && filterImage.Image != "" {
+			// Historically CRI-O has interpreted the “filter” as a single image to look up.
+			// Also, the type of the value is types.ImageSpec, the value used to refer to a single image.
+			// And, ultimately, Kubelet never uses the filter.
+			// So, fall back to existing code instead of having an extra code path doing some kind of filtering.
+			status, err := s.storageImageStatus(ctx, *filterImage)
+			if err != nil {
+				return nil, err
+			}
+			if status == nil {
+				return &types.ListImagesResponse{
+					Images: []*types.Image{},
+				}, nil
+			}
+			return &types.ListImagesResponse{
+				Images: []*types.Image{
+					ConvertImage(status),
+				},
+			}, nil
```

Comment:
we could update ConvertImage() to return an empty list instead of a single element, and then compress all of this to 
```
			return &types.ListImagesResponse{
				Images: ConvertImage(status),
			}, nil
```

Reply:
`ConvertImage` is also used below, in the primary `ListImages` cause, where there are almost always multiple images to collect in `Images`.

It might be cleaner to move all of the `filterImage.Image != "" ` code out of the main body, as a rare case into a separate function out of sight. OTOH “out of sight” might be a bad idea for a corner case — entirely up to CRI-O maintainers.

---

Or would you prefer
```go
resp := &types.ListImagesResponse{}
if status != nil {
  resp.Images = append(resp.Images, ConvertImage(status))
}
return resp, nil
```
here? I don’t have a preference either way.

Reply:
> as a rare case into a separate function out of sight.

I am open to this. Part of me feels like the code smells a bit that filters never really worked, and we probably should have them working in case kubelet uses them. 

> ConvertImage is also used below, in the primary ListImages cause, where there are almost always multiple images to collect in Images.

Another option is ConvertImages, which could take the filtered status here and then it could be called below. I am not feeling very convinced in any direction her--all of these are suggestions for consideration

Reply:
I've thought of it, I think I would prefer:

```
resp := &types.ListImagesResponse{}
if status != nil {
  resp.Images = append(resp.Images, ConvertImage(status))
}
return resp, nil
```

Reply:
Done.